Add docs for preventing volume mode conversion (#32673)
kubernetes/website: content/en/docs/concepts/storage/volume-snapshots.md

  driver: hostpath.csi.k8s.io
  source:
    volumeHandle: ee0cfb94-f8d4-11e9-b2d8-0242ac110002

>   sourceVolumeMode: Filesystem

Reviewer: It's possible that a user may be reading this doc but is using an older version of CRDs and controllers. So we need to add notes right next to this field to clarify how it can be used.

Author: Done!

Reviewer 2: @RaunakShah : This conversation was marked as resolved and I did not see @xing-yang 's comments actually addressed. Could you clarify if this was agreed upon to not be added eventually, please?

>   sourceVolumeMode: Filesystem # or Block. This field is only present for release-6.0 on VolumeSnapshotContent API.

Reviewer 3: > This field is only present for release-6.0 on VolumeSnapshotContent API.

If I use a later release, the text implies I can only use `Filesystem`. Am I right that it's SemVer and that 6.x and any later major version support the `Block` mode?
(we should make the text unambiguous; that also helps localization teams)

Reviewer: SourceVolumeMode is the mode of the volume whose snapshot is taken. It can be either “Filesystem” or “Block”. If not specified, it indicates the source volume's mode is unknown. This field is available in Volume Snapshot client v6.0.0 and higher.

Reviewer 3: @xing-yang what change should @RaunakShah make so that the situation is clear to readers?

Reviewer: I'm suggesting to add these https://github.com/kubernetes/website/pull/32673#discussion_r846635408 as comments next to the sourceVolumeMode.  Any other suggestions?

  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

`snapshotHandle` is the unique identifier of the volume snapshot created on the storage backend. This field is required for the pre-provisioned snapshots. It specifies the CSI snapshot id on the storage system that this `VolumeSnapshotContent` represents.

> `sourceVolumeMode` is the mode of the volume from which this snapshot was created. 

Reviewer: It's possible that a user may be reading this doc but is using an older version of CRDs and controllers.  So we need to add notes right next to this field to clarify how it can be used.

  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

`snapshotHandle` is the unique identifier of the volume snapshot created on the storage backend. This field is required for the pre-provisioned snapshots. It specifies the CSI snapshot id on the storage system that this `VolumeSnapshotContent` represents.
`sourceVolumeMode` is the mode of the volume from which this snapshot was created. 

## Converting the volume mode of a Snapshot

This feature is only present for `VolumeSnapshotContents` created with `v6.0.0` onwards. 

> For a user to create a `PersistentVolumeClaim` from an existing `VolumeSnapshot`, the annotation `snapshot.storage.kubernetes.io/allowVolumeModeChange: true` needs to be added to the `VolumeSnapshotContent` that corresponds to the `VolumeSnapshot`. An example is shown below.

Reviewer 3: ```suggestion
If you want to allow users to create a `PersistentVolumeClaim` from an existing `VolumeSnapshot`,
the annotation `snapshot.storage.kubernetes.io/allowVolumeModeChange: "true"` needs to be added
to the `VolumeSnapshotContent` that corresponds to the `VolumeSnapshot`. For example:
```

Notice that I made the annotation value be a string not a bool.

  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

`snapshotHandle` is the unique identifier of the volume snapshot created on the storage backend. This field is required for the pre-provisioned snapshots. It specifies the CSI snapshot id on the storage system that this `VolumeSnapshotContent` represents.

> `sourceVolumeMode` is the mode of the volume whose snapshot is taken. It can be either “Filesystem” or “Block”. If not specified, it indicates the source volume's mode is unknown. This field is available in Volume Snapshot client v6.0.0 and higher.

Reviewer 3: ```suggestion
`sourceVolumeMode` is the mode of the volume whose snapshot is taken. The value of the `sourceVolumeMode` field can be either `Filesystem` or `Block`. If you do not specify a source volume mode, Kubernetes treats the snapshot as if the source volume's mode is unknown. Support for this field is available in VolumeSnapshot client v6.0.0 and higher.
```

If someone asked me for the URL of the VolumeSnapshot client, I would not have a clue what that URL is or how to work it out. We should make it clear what we mean by the phrase “VolumeSnapshot client v6.0.0”.

Author: Thanks for the suggestion! I've used most of it but modified `If you do not specify a source volume mode` to `If the source volume mode is not specified`. It is not neccesarily on the user to populate this field.

> 

Reviewer: Add a link to https://github.com/kubernetes-csi/external-snapshotter/releases

Reviewer 2: I don't see this addressed either, ptal.

  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

`snapshotHandle` is the unique identifier of the volume snapshot created on the storage backend. This field is required for the pre-provisioned snapshots. It specifies the CSI snapshot id on the storage system that this `VolumeSnapshotContent` represents.

> `sourceVolumeMode` is the mode of the volume whose snapshot is taken. The value of the `sourceVolumeMode` field can be either `Filesystem` or `Block`. If the source volume mode is not specified, Kubernetes treats the snapshot as if the source volume's mode is unknown. Support for this field can only be enabled in VolumeSnapshot client v6.0.0 and higher. Visit the [VolumeSnapshot Release page](https://github.com/kubernetes-csi/external-snapshotter/releases) for more information.

Reviewer 3: Do you need to be using Kubernetes v1.24 to set this field?

Author: `VolumeSnapshot`s 6.0 release will only be supported with a minimum k8s version of 1.24. @xing-yang can correct me if I'm wrong here.

Reviewer 3: Do we mean:
> you need to be using `external-snapshotter` release 6.0 or later (which then requires Kubernetes v1.24)

?

Reviewer: VolumeSnapshot APIs are CRDs. So this field is not really in k/k code base.  
The only thing that is in k/k code base for the snapshot feature is the DataSource field in PVC which moved to GA in 1.20 release.
But we do update K8s dependencies in https://github.com/kubernetes-csi/external-snapshotter code and release it shortly after every K8s release. VolumeSnapshot client will be released shortly after K8s 1.24 release.

Reviewer 2: What we're trying to understand here is if there is a dependency for the version of `VolumeSnapshot`client on the Kubernetes release . You've mentioned that only the Datasource field (which moved to GA) is there in the k/k codebase. Are the other K8s dependencies that you mentioned related to other components within a release even though they're out of tree?

Reviewer: VolumeSnapshot client uses a few K8s components.  See here:
https://github.com/kubernetes-csi/external-snapshotter/blob/v6.0.0-rc4/client/go.mod#L6-L9
Currently it uses 1.23 components. After K8s 1.24 is released, the dependencies will be updated to 1.24 and we'll cut the official v6.0.0 client release.

> ## Converting the volume mode of a Snapshot {#convert-volume-mode}

Reviewer 3: nit:
```suggestion
## Converting the volume mode of a snapshot {#convert-volume-mode}
```

> This feature is only present for `VolumeSnapshotContents` created with client version `v6.0.0` onwards.

Reviewer 3: > This feature is only present for `VolumeSnapshotContents` created with client version `v6.0.0` onwards.

I'm guessing that a client can examine the `volumesnapshotcontents` API to see if it includes a `sourceVolumeMode` field; is that the only way for a client to tell whether the feature is supported?

Reviewer 2: Only speaking for myself here, but I think this would be good to elaborate @RaunakShah .

Reviewer: How about this: `This feature is in go module github.com/kubernetes-csi/external-snapshotter/client/v6 and later.`
I think a user can check if the new field `sourceVolumeMode` exists in a VolumeSnapshotContent spec or not by checking the manifest.

Reviewer 3: In these docs, we almost never talk about Go modules. Do end users need to look at that level of detail to make use of snapshots?

Imagine someone is doing development with a bit of operations, and their thoughts run like this:
- I have access to a Kubernetes cluster cluster
- There is a team who look after the platform, including that cluster
- I want to know whether the cluster lets me use `sourceVolumeMode`
- I know I can write a test to see if `sourceVolumeMode` works in my cluster and then use the result of that test
- I know I can ask the platform team to do the same test and let me know the outcome
- if `sourceVolumeMode` is available, I want to use it. But only in the development environment.

Reviewer 3: Can that developer write code that uses the Kubernetes API to work out whether `sourceVolumeMode` is available (eg by querying the cluster's OpenAPI data)?

If they can, let's give them a clear hint about that. Much better than to tell them to read the source code for a snapshotting component.
If the easiest way to work out what's available is to look at the Pod template for an add-on (`external-snapshotter`?) then we should tell them what to look for in that Pod template.

Reviewer: They can check if `sourceVolumeMode` is available in the manifest using kubectl: `kubectl get crd volumesnapshotcontent -o yaml`.
@RaunakShah Can you update the PR with this?

  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

`snapshotHandle` is the unique identifier of the volume snapshot created on the storage backend. This field is required for the pre-provisioned snapshots. It specifies the CSI snapshot id on the storage system that this `VolumeSnapshotContent` represents.

`sourceVolumeMode` is the mode of the volume whose snapshot is taken. The value 
of the `sourceVolumeMode` field can be either `Filesystem` or `Block`. If the 
source volume mode is not specified, Kubernetes treats the snapshot as if the 
source volume's mode is unknown. Support for this field can only be enabled in 

> `VolumeSnapshot` API v6.0.0 and higher. Visit the [VolumeSnapshot Release page](https://github.com/kubernetes-csi/external-snapshotter/releases) for more information.

Reviewer: Can you remove v6.0.0 as it is confusing for users?

## Converting the volume mode of a Snapshot {#convert-volume-mode}

> This feature is only present for `VolumeSnapshots` API with version `v6.0.0` onwards. To check if your 

Reviewer: Remove 6.0.0 here as well.

the mode of a volume.

To check if your cluster has capability for this feature, run the following command:

```yaml
$ kubectl get crd volumesnapshotcontent -o yaml

> ```

Reviewer 3: - this isn't YAML
- [don't include the command prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt)

> 

Reviewer 3: After running `kubectl get crd volumesnapshotcontent`, does a success exit indicate support for the  sourceVolumeMode` field? If not, we should document how to actually check.
